simulated_beacon/simulated_beacon.go

```diff
@@ -0,0 +1,240 @@
+type withdrawals struct {
+	pending []*types.Withdrawal
+	next    uint64
+	mu      sync.Mutex
+}
+
+// pop removes up to 10 withdrawals from the queue
+func (w *withdrawals) pop() []*types.Withdrawal {
+	w.mu.Lock()
+	defer w.mu.Unlock()
+
+	var popCount int
+	if len(w.pending) >= 10 {
+		popCount = 10
+	} else {
+		popCount = len(w.pending)
+	}
+
+	popped := make([]*types.Withdrawal, popCount)
+	copy(popped[:], w.pending[0:popCount])
+	w.pending = append([]*types.Withdrawal{}, w.pending[popCount:]...)
+	return popped
+}
+
+// add adds a withdrawal to the queue
+func (w *withdrawals) add(withdrawal *types.Withdrawal) error {
+	w.mu.Lock()
+	defer w.mu.Unlock()
+
+	if withdrawal.Index < w.next {
+		return fmt.Errorf("withdrawal has index (%d) less than or equal to latest received withdrawal index (%d)", withdrawal.Index, w.next-1)
+	}
+	w.next = withdrawal.Index + 1
+	w.pending = append(w.pending, withdrawal)
+	return nil
+}
+
+type SimulatedBeacon struct {
+	shutdownCh   chan struct{}
+	eth          *eth.Ethereum
+	period       time.Duration
+	withdrawals  withdrawals
+	feeRecipient common.Address
+	// mu gates concurrent access to the feeRecipient
+	mu sync.Mutex
+}
+
+func NewSimulatedBeacon(eth *eth.Ethereum) *SimulatedBeacon {
+	chainConfig := eth.APIBackend.ChainConfig()
+	if chainConfig.Dev == nil {
+		log.Crit("incompatible pre-existing chain configuration")
+	}
+
+	return &SimulatedBeacon{
+		eth:          eth,
+		period:       time.Duration(chainConfig.Dev.Period) * time.Second,
+		withdrawals:  withdrawals{[]*types.Withdrawal{}, 0, sync.Mutex{}},
+		feeRecipient: common.Address{},
+	}
+}
+
+func (c *SimulatedBeacon) setFeeRecipient(feeRecipient *common.Address) {
+	c.mu.Lock()
+	c.feeRecipient = *feeRecipient
+	c.mu.Unlock()
+}
+
+func (c *SimulatedBeacon) getFeeRecipient() common.Address {
+	c.mu.Lock()
+	feeRecipient := c.feeRecipient
+	c.mu.Unlock()
+
+	return feeRecipient
+}
+
+// Start invokes the SimulatedBeacon life-cycle function in a goroutine
+func (c *SimulatedBeacon) Start() error {
+	c.shutdownCh = make(chan struct{})
+	go c.loop()
+	return nil
+}
+
+// Stop halts the SimulatedBeacon service
+func (c *SimulatedBeacon) Stop() error {
+	close(c.shutdownCh)
+	return nil
+}
+
+// loop manages the lifecycle of the SimulatedBeacon.
+// it drives block production, taking the role of a CL client and interacting with Geth via public engine/eth APIs
+func (c *SimulatedBeacon) loop() {
+	var (
+		ticker             = time.NewTicker(time.Millisecond * 100)
+		lastBlockTime      = time.Now()
+		engineAPI          = catalyst.NewConsensusAPI(c.eth)
+		header             = c.eth.BlockChain().CurrentHeader()
+		curForkchoiceState = engine.ForkchoiceStateV1{
+			HeadBlockHash:      header.Hash(),
+			SafeBlockHash:      header.Hash(),
+			FinalizedBlockHash: header.Hash(),
+		}
+	)
+
+	// if genesis block, send forkchoiceUpdated to trigger transition to PoS
+	if header.Number.BitLen() == 0 {
```

Comment:
```suggestion
	if header.Number.Sign() == 0 {
```
A minor nit, which is of no consequence in this case, but generally, `Sign` is better than `BitLen`, since sign just checks the length of `x.abs` whereas `BitLen` actually iterates over the `x.abs`